### PR TITLE
common, osd: optimize OpTracker and request enqueue path

### PR DESCRIPTION
**Pull request by rzarzynski**
Summary
=======
This is a continuation of PR #20702. It includes almost all its changes and enriches it further with additional optimizations related to the request creation and destruction phases.
 
The avoidance of unnecessary `TrackedOp::nref` toggling, the main idea behind the missing commit, has been incorporated separately as a fix for racy `OpTracker::visit_ops_in_flight` in PR #22877.

Notable changes since the previous PR
---------------------------------------------------
* Critical sections of the `Throttle`, present on the hot path, has been significantly minimized or avoided. The responsible commits have been sent for an early review as PR #22605.
* Selecting of `OpTracker`'s shard is conducted basing on `spg_t` if possible. This establishes a relationship with respective shards of the main work queue and avoids cache line ping-pong between messenger threads.
* `OpTracker::create_request()` has been optimized. It skips a dynamic memory allocation and chronic mutex reacquisition.
* Cheaper memory ordering model for `TrackedOp::state` has been used to avoid `MFENCE`.
* Unnecessary ref-counting has been removed from `OSD::enqueue_op`.
*  `ceph::math::p2_t` has been integrated with Ceph's config subsystem allowing the `isp2` validation to be performed early and once -- at the config parsing.

History
---------
**UPDATE** July 31th 2018: updated summary.
**UPDATE** July 26th 2018: added numbers useful when comparing with the old `OpTracker` PR.
**UPDATE** July 27th 2018: added numbers from `vstart.sh`-based testing.

Performance comparison
===================
All tests have been conducted on the `incerta07` node in 5 rounds. The result for each round is sum of bandwidth 4 separated FIO-librbd clients get from a cluster consisted of 3 OSDs . CBT has been used for deployment and orchestration. Values are in KBps.

On average for 4 KiB randreads:
```
>>> sum([408676,  406235,  411658,  397742,  400255.])/sum([377443,  380011,  382396,  375046,  382870])
1.066815402952735
```
On average for 4 KiB randreads (these changes + PR #22877; see below)
```
>>> sum([408676,  406235,  411658,  397742,  400255.])/sum([376844,  375004,  375534,  382742,  378275])
1.072107112956531
```
The reference point`~`: 910f3aec1ac81ea1641a2ca714824b97af5c92ef
--------------------------------------------------------------
For comparison with [the previous OpTracker PR](https://github.com/ceph/ceph/pull/22877) it might be reasonable to start from commit 910f3aec1ac81ea1641a2ca714824b97af5c92ef. It directly precedes 360211ed7b8b0c3a87432ae13e189a51a1e9a026 that incorporated into `master` an idea/optimization that was a part of the original rework -- it appeared to fix racing `visit_ops_in_flight()`.

```
randread,  4 KiB:  376844,  375004,  375534,  382742,  378275
randread, 64 KiB: 3565172, 3554422, 3569961, 3354016, 3638335
```

The reference point: 360211ed7b8b0c3a87432ae13e189a51a1e9a026
-------------------------------------------------------------
```
randread,  4 KiB:  377443,  380011,  382396,  375046,  382870
randread, 64 KiB: 3635133, 3580268, 3568959, 3540768, 3558203
```

After the changes: fefd4a79871301a2818a74374d9af44eb4b97256
-----------------------------------------------------------
```
randread,  4 KiB:  408676,  406235,  411658,  397742,  400255
randread, 64 KiB: 3688354, 3754750, 3634474, 3572501, 3669665
```

Performance comparison with `vstart.sh`
=============================================
All tests have been conducted on the `incerta07` node in 5 rounds. The result for each round is sum of bandwidth 4 separated FIO-librbd clients get from a cluster consisted of 3 OSDs and deployed by `vstart.sh --nolockdep`. Deployment and orchestration was automated with a dumb, self-written bash script, so please be careful -- personally I put much more trust in CBT. Values are in KBps, huge contention observed.

On average for 4 KiB randreads:
```
>>> sum([94596, 93283, 91940, 87585, 95401])/sum([86857, 80899, 85900, 86139, 88868.])
1.0796476486190785
```

The reference point: 360211ed7b8b0c3a87432ae13e189a51a1e9a026
-------------------------------------------------------------
```
randread, 4 KiB:  86857, 80899, 85900, 86139, 88868
```

After the changes: fefd4a79871301a2818a74374d9af44eb4b97256
-----------------------------------------------------------
```
randread, 4 KiB:  94596, 93283, 91940, 87585, 95401
```

Profiling in the domain of `cycles:pp`
=========================
4 KB randreads, `master`, the same machine like in the performance comparison.
```
Samples: 495K of event 'cycles:pp', Event count (approx.): 152898961741
  Children      Self  Command          Shared Object         Symbol                                                                                                                                          ◆
+   62,09%     0,00%  tp_osd_tp        ceph-osd              [.] ShardedThreadPool::WorkThreadSharded::entry                                                                                                 ▒
+   62,09%     0,00%  tp_osd_tp        libpthread-2.17.so    [.] start_thread                                                                                                                                ▒
+   62,09%     0,00%  tp_osd_tp        libc-2.17.so          [.] __clone                                                                                                                                     ▒
+   62,09%     0,10%  tp_osd_tp        ceph-osd              [.] ShardedThreadPool::shardedthreadpool_worker                                                                                                 ▒
-   61,88%     0,67%  tp_osd_tp        ceph-osd              [.] OSD::ShardedOpWQ::_process                                                                                                                  ▒
   - 61,21% OSD::ShardedOpWQ::_process                                                                                                                                                                       ▒
      + 40,46% PGOpItem::run                                                                                                                                                                                 ▒
      - 13,16% PGOpItem::~PGOpItem                                                                                                                                                                           ▒
         - TrackedOp::put                                                                                                                                                                                    ▒
            - 12,24% OpTracker::record_history_op                                                                                                                                                            ▒
                 0,09% pthread_rwlock_rdlock                                                                                                                                                                 ▒
                 0,04% std::__detail::_List_node_base::_M_hook                                                                                                                                               ▒
                 0,03% pthread_rwlock_unlock                                                                                                                                                                 ▒
                 0,02% operator new
```

```
OpTracker::record_history_op  /usr/local/bin/ceph-osd
...
  0,08 │ e0:   mov    %ecx,%edx
 97,23 │       xchg   %dl,(%rax)
       │     _ZN4ceph11version_1_09spin_lockERSt11atomic_flag():
       │     };
       │
```

### CODE REVIEW

**Comment by mattbenjamin** [comment]
Don't know about Piotr's concerns, but the whole looks valuable.  Would be worth documenting so it can move forward.

File: src/common/TrackedOp.cc [hunk 1]
```diff
@@ -256,18 +256,14 @@ void OpTracker::register_inflight_op(TrackedOp *i,
...
-    i->state = TrackedOp::STATE_LIVE;
+    i->state.store(TrackedOp::STATE_LIVE, std::memory_order_release);
   }
 }
 
 void OpTracker::unregister_inflight_op(TrackedOp* const i)
 {
   // caller checks;
-  assert(i->state);
+  assert(i->state.load(std::memory_order_acquire));
```
Comment (mattbenjamin) [inline src/common/TrackedOp.cc]: after all this, do we want asserts?

**Comment by branch-predictor** [comment]
@mattbenjamin @gregsfortytwo My concerns were mostly related to the fact that entire PR could be split into several PRs, as they're affecting Ceph in general and not just OpTracker (see https://github.com/ceph/ceph/pull/23071/commits/6792da9f1e8d666bedf2922b4bfdc08365391b11 or https://github.com/ceph/ceph/pull/23071/commits/f5ef947c9c0af676c9f59642202d9f626474f495 for examples). That's still the case.
Also, I'm still not convinced about optimizing divisions, as these are integer divisions and nowhere close in cost to floating point divisions - sounds like optimizing just for sake of optimizing at the cost of extra code (see https://github.com/ceph/ceph/pull/23071/commits/6792da9f1e8d666bedf2922b4bfdc08365391b11 for example).

**Comment by rzarzynski** [comment]
@gregsfortytwo, @mattbenjamin, @branch-predictor: The PR doesn't include all commits from the previous PR #20702 as the idea to avoid unnecessary ref-counting in `OpTracker::put` turned into a bug fix and has been recently merged. The merge commit became the starting point of this branch. To make comparison easier I updated description with additional results.

I personally perceive the changes as a whole as substantial part of the benefit comes from aggregating large number of small changes. For instance, the group touching `Throttle` gives ~1,06%, the `put()` fix/optimization brought ~0,49% while the commits eradicating `division-memory indirection-memory-indirection` are responsible for ~0.26%. However, it's hard to measure such small differences reliably and they can interfere.

Judging the right threshold of gain-to-churn ratio belongs to reviewer. What I would like to emphasize is just for many of the things proposed to `common` the `OpTracker` is an initial user. I believe making them in reusable manner can be a response for similar problems across other parts of the code.

> they're affecting Ceph in general and not just OpTracker (see 6792da9 or f5ef947 for examples). That's still the case.

Agreed. Keeping the spinlock optimization has been described in the summary of PR #20702.

> Also, I'm still not convinced about optimizing divisions, as these are integer divisions and nowhere close in cost to floating point divisions - sounds like optimizing just for sake of optimizing at the cost of extra code (see 6792da9 for example).

Piotr, didn't you say that "*[`ceph_clock_now()`] is asking glibc to return current real time - which by itself is slow*" and "*That makes it impossible for division you optimized to take 35% of cpu time because it's not ~2 times faster than `clock_gettime`, but about 10 times faster*"? Now we know that single 64 bit integer division (2 have been eradicated) can take 30-90 cycles while [the profiling shows the function requires on average ~68 cycles](https://gist.github.com/rzarzynski/525c7eb09e5802521cd140ce424fe4c1). This would sum up to an equivalent of the call you consider costly.

However, all of this comes just from in-vitro testing and measuring such tiny differences in-vivo is very hard, so let me repeat: *killing couple of divisions is a tiny optimization at best*. With that being said, I still consider `p2_t` as a valuable, generic bit because:

 * it deals with assertions like we have in e.g. BlueStore ([1](https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/BlueStore.cc#L4467), [2](https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/BitmapFreelistManager.cc#L65), [3](https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/BlueStore.h#L2823), [4](https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/fastbmap_allocator_impl.h#L616), ...). As the information about p2-iness is embedded directly in the type, programmers are free of ensuring the invariant manually.
 * The validation can be performed very early - on config parsing. This minimizes chances for failing an assert buried deeply on an infrequent path. It also kills overhead from having multiple, duplicative asserts on hot path.
 * Divisions and modulo are free of bit-twiddling that IMHO appears very messy and [sometimes impacts even an interface (hot paths in BlueFS)](https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/bluefs_types.h#L131). Overloaded operators still behave in the intuitive manner. Conceptually they aren't different from what compilers do to not emit `DIV`.

> these are integer divisions and nowhere close in cost to floating point divisions

I believe this needs an quotation as it appears IMHO contrary to [Agner Fog's instruction tables](https://www.agner.org/optimize/instruction_tables.pdf) for Haswell.

Integer instructions of Haswell, p. 201

| Instruction | Operands | μops fused domain | μops unfused domain | μops each port | Latency | Recipro-cal throughput |
| ------------| ---------| ------------------|---------------------| ---------------|---------|------------------------|
| DIV | r8   | 9  | 9  | p0 p1 p5 p6 | 22-25 | 9 |
| DIV | r16 | 11 | 11 | p0 p1 p5 p6 | 23-26 | 9 |
| DIV | r32 | 10 | 10 | p0 p1 p5 p6 | 22-29 | 9-11 |
| DIV | r64 | 36 | 36 | p0 p1 p5 p6 | 32-96 | 21-74 |

Floating point x87 instructions of Haswell, p. 205

| Instruction | Operands | μops fused domain | μops unfused domain | μops each port | Latency | Recipro-cal throughput |
| ------------| ---------| ------------------|---------------------| ---------------|---------|------------------------|
| FDIV(R)(P) | r | 1 | 1 | p0 | 10-24 | 8-18 |

**Comment by branch-predictor** [comment]
On 18-07-31 12:36 AM, Radoslaw Zarzynski wrote:
> 
>     Also, I'm still not convinced about optimizing divisions, as these are
>     integer divisions and nowhere close in cost to floating point divisions
>     - sounds like optimizing just for sake of optimizing at the cost of
>     extra code (see 6792da9
>     <https://github.com/ceph/ceph/commit/6792da9f1e8d666bedf2922b4bfdc08365391b11>
>     for example).
> 
> Piotr, didn't you say that "/[|ceph_clock_now()|] is asking glibc to return 
> current real time - which by itself is slow/" and "/That makes it impossible 
> for division you optimized to take 35% of cpu time because it's not ~2 times 
> faster than |clock_gettime|, but about 10 times faster/"? Now we know that 
> single 64 bit integer division (2 have been eradicated) can take 30-90 
> cycles while the profiling shows the function requires on average ~68 cycles 
> <https://gist.github.com/rzarzynski/525c7eb09e5802521cd140ce424fe4c1>. This 
> would sum up to an equivalent of the call you consider costly.

Yes, getting rid of unnecessary calls to ceph_clock_now() is always a good 
idea and it's easy to do in most of cases.

> However, all of this comes just from in-vitro testing and measuring such 
> tiny differences in-vivo is very hard, so let me repeat: /killing couple of 
> divisions is a tiny optimization at best/. With that being said, I still 
> consider |p2_t| as a valuable, generic bit because:
> 
>   * it deals with assertions like we have in e.g. BlueStore (1
>     <https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/BlueStore.cc#L4467>,
>     2
>     <https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/BitmapFreelistManager.cc#L65>,
>     3
>     <https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/BlueStore.h#L2823>,
>     4
>     <https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/fastbmap_allocator_impl.h#L616>,
>     ...). As the information about p2-iness is embedded directly in the
>     type, programmers are free of ensuring the invariant manually.
>   * The validation can be performed very early - on config parsing. This
>     minimizes chances for failing an assert buried deeply on an infrequent
>     path. It also kills overhead from having multiple, duplicative asserts
>     on hot path.
>   * Divisions and modulo are free of bit-twiddling that IMHO appears very
>     messy and sometimes impacts even an interface (hot paths in BlueFS)
>     <https://github.com/rzarzynski/ceph/blob/fefd4a79871301a2818a74374d9af44eb4b97256/src/os/bluestore/bluefs_types.h#L131>.
>     Overloaded operators still behave in the intuitive manner. Conceptually
>     they aren't different from what compilers do to not emit |DIV|.

If you'd really mean it, you'd just ditch p2_t-related changes to separate 
PR that would deal with these issues in entire Ceph code base, whole-sale.

>     these are integer divisions and nowhere close in cost to floating point
>     divisions
> 
> I believe this needs an quotation as it appears IMHO contrary to Agner Fog's 
> instruction tables <https://www.agner.org/optimize/instruction_tables.pdf> 
> for Haswell.
> 
> Integer instructions of Haswell, p. 201
> 
> Instruction 	Operands 	μops fused domain 	μops unfused domain 	μops each 
> port 	Latency 	Recipro-cal throughput
> DIV 	r8 	9 	9 	p0 p1 p5 p6 	22-25 	9
> DIV 	r16 	11 	11 	p0 p1 p5 p6 	23-26 	9
> DIV 	r32 	10 	10 	p0 p1 p5 p6 	22-29 	9-11
> DIV 	r64 	36 	36 	p0 p1 p5 p6 	32-96 	21-74
> 
> Floating point x87 instructions of Haswell, p. 205
> 
> Instruction 	Operands 	μops fused domain 	μops unfused domain 	μops each 
> port 	Latency 	Recipro-cal throughput
> FDIV(R)(P) 	r 	1 	1 	p0 	10-24 	8-18

You're forgetting that FDIV results are on FPU stack and need to be moved 
from there with separate instructions that also take μops on both cpu and 
fpu (and clutter the instruction pipeline in the process).

-- 
Piotr Dałek
piotr.dalek@corp.ovh.com
https://www.ovhcloud.com

**Pull request closed without merging**